# Question 1: Dataset Overview
---

- Begin by loading the dataset and using summary functions to gain an initial understanding of the variables, their types, and the presence of any missing data. Discuss the importance of assessing data structures in the context of statistical modeling.

In [ ]:
# Step 0: Install the necessary packages and load them
---

In [36]:
install.packages("dplyr")


The downloaded binary packages are in
	/var/folders/nv/sjj_9gb52674c8ktybqghkm80000gq/T//Rtmp0rDFXY/downloaded_packages


In [37]:
library(dplyr)

## Step 1: Load the dataset
---

In [39]:
load("Df_regression.RData")

## Step 2: Check the content of the data
---

In [40]:
ls()

[1] "Df_regression_unique" "dataset"

## Step 3: Check the class of the data
---

In [41]:
class(Df_regression_unique)
class(dataset)

[1] "data.frame"

[1] "character"

### Explanation:
---
1. Df_regression_unique is the 'dataframe' while dataset contains the 'character'.
2. It is obvious then that Df_regression_unique contains the data.

## Step 4: Get a snapshot of the data
---
1. Snapshots are important because we can see how our data looks like without looking at the entire dataset.

2. In my practice I usually look at the tail in of the dataset, for me this is ideal because of two things:

    a. By using the tail we can check how many rows the dataset is.

    b. We can also check the datatype of each columns.


In [42]:
tail(Df_regression_unique)

ptnum label C-424144002 C-263495000 C-103579009 C-72166-2 C-763302001
5933 p9972 0     53          f           white       never     TRUE       
5934 p9976 0     70          m           white       former    TRUE       
5935 p9982 0     70          m           black       never     TRUE       
5936 p9992 1     75          m           white       never     TRUE       
5937 p9996 0     59          m           white       never     TRUE       
5938 p9998 0     46          f           white       former    TRUE       
     C-39156-5 C-185086009 C-21908-9 C-233604007 C-195967001 C-132281000119108
5933 29.66     NA          NA        NA          NA          NA               
5934 30.06     NA          NA        NA          NA          NA               
5935 31.07     NA          NA        NA          NA          NA               
5936 29.60     NA          NA        NA          NA          NA               
5937 27.30     NA          NA        NA          NA          NA               
5938 27.60     NA          NA        NA          NA          NA               
     Followup Exposure
5933  822     1       
5934 3042     1       
5935 1233     1       
5936 1838     1       
5937 2374     2       
5938  356     1

## Step 5: Counting the number of nulls in the columns.
---
1. In the previous cell we have observed that columns: C-1850860095796, C-21908-95715C-2336040075754, C-1959670015932, and C-1322810001191085880 contains null.

2. Strategy: 

    a. We know that the total number of columns is 5,938, we need to compare how many nulls are in these columns.
    b. If the nulls are greater than 50% that is:
    
    $\text{% null} = \frac{\text{total number of nulls in a column}}{\text{total number of rows}}$
    
    then it is better to drop the columns with more than 50% missing than spending effort on testing if the nulls are missing by random.

In [43]:
colSums(is.na(Df_regression_unique))

ptnum             label       C-424144002       C-263495000 
                0                 0                 0                 0 
      C-103579009         C-72166-2       C-763302001         C-39156-5 
                0                 0              1020                 0 
      C-185086009         C-21908-9       C-233604007       C-195967001 
             5796              5715              5754              5932 
C-132281000119108          Followup          Exposure 
             5880                 0                 0

## Step 6: Remove columns with more than 50% missing values.
---
Note: Make sure the dplyr library is installed here.

If the column names contain special characters (like -), the dplyr package is a more robust solution. Use backticks ("<name column here>") to reference such column names.

In [49]:
Df_regression_unique <- Df_regression_unique %>% select(-"C-185086009", -"C-21908-9", -"C-233604007", -"C-195967001", -"C-132281000119108")

## Step 7: Check is the columns are properly removed.
---

In [50]:
tail(Df_regression_unique)

ptnum label C-424144002 C-263495000 C-103579009 C-72166-2 C-763302001
5933 p9972 0     53          f           white       never     TRUE       
5934 p9976 0     70          m           white       former    TRUE       
5935 p9982 0     70          m           black       never     TRUE       
5936 p9992 1     75          m           white       never     TRUE       
5937 p9996 0     59          m           white       never     TRUE       
5938 p9998 0     46          f           white       former    TRUE       
     C-39156-5 Followup Exposure
5933 29.66      822     1       
5934 30.06     3042     1       
5935 31.07     1233     1       
5936 29.60     1838     1       
5937 27.30     2374     2       
5938 27.60      356     1

## Discussion:
---
Assessing data structures is a crucial step in preparing for statistical modeling. It ensures that the data is correctly understood, processed, and interpreted, ultimately leading to valid and reliable results. Here’s why it’s important:

1. Understanding Data Types
Significance:
Each variable type (e.g., numeric, categorical, logical) dictates the kinds of analyses and transformations that can be performed.
For example, regression models treat numeric and categorical variables differently.
Potential Issues:
A categorical variable coded as numeric may lead to incorrect interpretation (e.g., treating regions as continuous numbers).
Logical variables not converted to binary (0/1) may confuse algorithms.
Solution:
Use functions like str() and summary() to ensure variables are properly typed.


2. Detecting and Addressing Missing Data
Significance:
Missing data can skew model results, reduce statistical power, and introduce biases.
Understanding patterns of missingness (e.g., Missing Completely at Random, Missing Not at Random) informs appropriate imputation methods.
Potential Issues:
Ignoring missing data can result in smaller, unrepresentative datasets if rows with missing values are removed.
Arbitrary imputation can distort relationships in the data.
Solution:
Visualize missing data patterns (naniar in R) and assess whether imputation, removal, or advanced techniques (e.g., multiple imputation) are needed.


3. Identifying Outliers and Inconsistent Values
Significance:
Outliers can have a disproportionate influence on statistical models, especially in regression and clustering.
Inconsistent values (e.g., negative ages, nonsensical categories) can indicate data entry errors.
Potential Issues:
Models may overfit or misinterpret trends due to extreme or incorrect values.
Solution:
Perform exploratory data analysis (EDA) to detect and decide how to handle outliers.


4. Ensuring Data Conforms to Model Requirements
Significance:
Many statistical models have assumptions (e.g., linear regression assumes linearity, normality, homoscedasticity).
Proper formatting (e.g., encoding categorical variables as factors) ensures compatibility with modeling functions.
Potential Issues:
Non-normal data in a model requiring normality can lead to invalid inferences.
Misformatted data might cause errors or incorrect results in statistical software.
Solution:
Check distributions, relationships, and data transformations (e.g., log transformations for skewed data).


5. Reducing Dimensionality and Improving Interpretability
Significance:
High-dimensional data can cause overfitting and make results harder to interpret.
Dimensionality reduction techniques (e.g., PCA, feature selection) require an understanding of variable roles.
Potential Issues:
Irrelevant variables add noise and reduce the efficiency of the model.
Solution:
Assess variable relevance and multicollinearity before modeling.


6. Recognizing Data Relationships
Significance:
Relationships between variables (e.g., collinearity, interactions) influence model complexity and performance.
Potential Issues:
Ignoring relationships can lead to misspecified models or incorrect interpretations.
Solution:
Explore relationships using scatterplots, correlation matrices, or advanced methods like variance inflation factors (VIF).


7. Ensuring Ethical and Accurate Representation
Significance:
Assessing data structures ensures that the dataset represents the population accurately.
Potential Issues:
Imbalanced datasets can bias models (e.g., under-representation of minority groups in healthcare studies).
Solution:
Stratify data, balance classes, and ensure fair sampling.

# Question 2: Handling Missing Data
---
Identify columns with logical (TRUE/FALSE) values and apply methods for addressing missing values in these columns, replacing them with 0 or 1 based on their logical value. Reflect on the implications of missing data and the
strategies used for handling it.

## Step 0: Direct Conversion in Base R
---

Logical to Integer: In R, logical values are internally represented as TRUE = 1 and FALSE = 0. The as.integer() function converts these directly.

In [52]:
Df_regression_unique$"C-763302001" <- as.integer(Df_regression_unique$"C-763302001")

## Step 1: Verify the Changes
----

In [55]:
tail(Df_regression_unique)

ptnum label C-424144002 C-263495000 C-103579009 C-72166-2 C-763302001
5933 p9972 0     53          f           white       never     1          
5934 p9976 0     70          m           white       former    1          
5935 p9982 0     70          m           black       never     1          
5936 p9992 1     75          m           white       never     1          
5937 p9996 0     59          m           white       never     1          
5938 p9998 0     46          f           white       former    1          
     C-39156-5 Followup Exposure
5933 29.66      822     1       
5934 30.06     3042     1       
5935 31.07     1233     1       
5936 29.60     1838     1       
5937 27.30     2374     2       
5938 27.60      356     1

## Step2:  If There Are NA Values
---

If the column contains (NA) values, decide how to handle them. For instance, you might replace them with 0.

In [58]:
Df_regression_unique$"C-763302001"[is.na(Df_regression_unique$"C-763302001")] <- 0
Df_regression_unique$"C-763302001" <- as.integer(Df_regression_unique$"C-763302001")

In [59]:
tail(Df_regression_unique)

ptnum label C-424144002 C-263495000 C-103579009 C-72166-2 C-763302001
5933 p9972 0     53          f           white       never     1          
5934 p9976 0     70          m           white       former    1          
5935 p9982 0     70          m           black       never     1          
5936 p9992 1     75          m           white       never     1          
5937 p9996 0     59          m           white       never     1          
5938 p9998 0     46          f           white       former    1          
     C-39156-5 Followup Exposure
5933 29.66      822     1       
5934 30.06     3042     1       
5935 31.07     1233     1       
5936 29.60     1838     1       
5937 27.30     2374     2       
5938 27.60      356     1

## Discussion
---

Missing data is a common challenge in statistical modeling and data analysis. Its presence can significantly impact the validity, reliability, and interpretability of results. Proper handling of missing data is crucial to ensure robust and unbiased conclusions.

Implications of Missing Data:

1. Reduced Statistical Power
Missing data reduces the sample size available for analysis, weakening the statistical power of tests.
Smaller datasets may lead to inconclusive results or increased Type II errors (failing to detect true effects).

2. Bias in Results
Missing data that is not random (e.g., people with certain traits systematically omit answers) can bias estimates.
For example, in medical studies, patients with severe symptoms may drop out, leading to underestimation of the disease severity.


3. Distorted Relationships
Missing data can change the relationships between variables if handled improperly.
For instance, imprecise imputation might introduce spurious correlations.


4. Loss of Representativeness
If a particular subgroup is more likely to have missing data, the remaining data may no longer represent the entire population.


5. Impact on Model Performance
Machine learning and statistical models often fail or perform poorly with missing data unless explicitly accounted for.


## Strategies for Handling Missing Data
---

1. Prevention

    Careful Data Collection: Design surveys and data collection processes to minimize missing values.
    
    Data Validation: Use real-time checks during data entry to catch missing or incorrect inputs.

<br>
2. Understanding the Missingness Mechanism

    Missing Completely at Random (MCAR): Data is missing with no systematic relationship to the observed or unobserved data. This is the least problematic type.

    Missing at Random (MAR): The likelihood of missingness depends on observed variables but not unobserved ones.

    Missing Not at Random (MNAR): Missingness depends on unobserved variables. This requires specialized techniques or assumptions.

<br>
3. Strategies to Handle Missing Data

    A. Deletion Methods
        Listwise Deletion: Remove rows with any missing values.
        
        Advantages: Simple and retains consistency across analyses.
        
        Disadvantages: Reduces sample size, can bias results if data is not MCAR.
        
        Pairwise Deletion: Use available data for each analysis without discarding entire rows.
        
        Advantages: Retains more data.
        
        Disadvantages: Results may vary across analyses.
        

    B. Imputation Methods
        Mean/Median/Mode Imputation: Replace missing values with the mean (for numeric) or mode (for categorical).
        Advantages: Simple and quick.
        Disadvantages: Reduces variability, can introduce bias.
        Regression Imputation: Use regression models to predict and fill missing values based on other variables.
        Advantages: Accounts for relationships between variables.
        Disadvantages: Assumes the imputation model is correct, underestimates variability.
        Multiple Imputation: Create multiple plausible datasets by imputing missing values with different estimates and combining results.
        Advantages: Accounts for uncertainty and variability in imputations.
        Disadvantages: Computationally intensive, complex to implement.
        Hot Deck Imputation:Replace missing values with observed values from similar cases.
        Advantages: Retains observed data distribution.
        Disadvantages: Depends on finding similar cases.

    C. Model-Based Approaches
        Maximum Likelihood Estimation (MLE): Estimate parameters directly while accounting for missing data.
        Advantages: Efficient and unbiased under MAR.
        Disadvantages: Requires complex algorithms.
        Bayesian Methods: Use prior distributions to estimate missing values.
        Advantages: Flexible and accounts for uncertainty.
        Disadvantages: Requires expertise and computational power.

    D. Advanced Techniques
        K-Nearest Neighbors (KNN): Impute missing values based on the closest neighbors in the dataset.
        Advantages: Considers the data's structure.
        Disadvantages: Computationally expensive for large datasets.
        Machine Learning: Use predictive models like Random Forest to impute missing values.
        Advantages: Handles complex relationships.
        Disadvantages: May overfit and requires validation.

# Question 3: Categorical Variable Conversion
---

Convert relevant variables into categorical (factor) variables as necessary, and consider the role of this transformation in regression analyses. Discuss when and why categorical conversion is crucial in epidemiological research.

In [ ]:




# Importance of assessing data structures
cat("Understanding data types and missing values helps ensure the dataset is suitable for regression analyses.\n")

# 2. Handling Missing Data in Logical Columns
logical_cols <- sapply(dataset, is.logical)
logical_missing <- which(colSums(is.na(dataset[, logical_cols])) > 0)

# Replace NA in logical columns with 0 or 1 based on logical value
dataset[, logical_cols] <- lapply(dataset[, logical_cols], function(col) {
  ifelse(is.na(col), 0, as.integer(col))
})

cat("Logical missing values replaced.\n")

# 3. Categorical Variable Conversion
# Identify potential categorical columns (e.g., character or specific variables)
categorical_vars <- sapply(dataset, is.character)

# Convert to factors
dataset[, categorical_vars] <- lapply(dataset[, categorical_vars], as.factor)

cat("Character variables converted to factors.\n")

# 4. Data Cleaning: Handling Missing in Categorical Variables
# Example: Impute missing stages with "Unknown" (adjust column name as needed)
if ("disease_stage" %in% colnames(dataset)) {
  dataset$disease_stage[is.na(dataset$disease_stage)] <- "Unknown"
  dataset$disease_stage <- as.factor(dataset$disease_stage)
}

cat("Categorical missing values addressed.\n")

# 5. Missing Data Pattern Visualization
library(ggplot2)
library(naniar)

# Visualize missing data pattern
gg_miss_var(dataset, show_pct = TRUE)

cat("Missing data pattern visualized.\n")

# 6. Descriptive Statistics and Summary Tables
library(dplyr)
library(tableone)

# Example: Create summary table stratified by an exposure group (e.g., `exposure`)
if ("exposure" %in% colnames(dataset)) {
  summary_table <- CreateTableOne(vars = colnames(dataset), strata = "exposure", data = dataset)
  print(summary_table)
} else {
  cat("Exposure variable not found; summary generated without stratification.\n")
  print(summary(dataset))
}

cat("Descriptive statistics completed.\n")
